liquid-compiler/src/parser.rs

```diff
@@ -175,14 +176,39 @@ fn parse_value(value: Pair) -> Expression {
 
 /// Parses a `FilterCall` from a `Pair` with a filter.
 /// This `Pair` must be `Rule::Filter`.
-fn parse_filter(filter: Pair, options: &Language) -> Result<FilterCall> {
+fn parse_filter(filter: Pair, options: &Language) -> Result<Box<Filter>> {
     if filter.as_rule() != Rule::Filter {
         panic!("Expected a filter.");
     }
 
     let mut filter = filter.into_inner();
     let name = filter.next().expect("A filter always has a name.").as_str();
-    let args = filter.map(parse_value).collect();
+
+    let mut keyword_args = HashMap::new();
+    let mut positional_args = Vec::new();
```

Comment:
Alternatives: 
- `Vec<(Option<&str>, &str)>`
- `(Vec<&str>, Vec<&str, &str>)`

I'm not sure what having positional arguments after keyword arguments should do.  Error? Actually work with ordering being important?  It might be useful for us to preserve the order of keyword or all arguments. 

I'm also assuming we'll just iterate over these rather than doing lookups, which would be better tuned for `Vec`.

Thoughts?

Reply:
In my opinion, the keyword arguments should just be skipped in the positional count (so in `filter: 1, a:2, 3`, 3 would be the 2nd positional arg).

As for the alternatives, wouldn't that make it hard to lookup the keyword arguments?

Another thing that I've been thinking is: is it really needed to have both positional and keyword args in the same filter? Is there any liquid/jekyll/other filter that does that? Maybe we could just allow either positional or keyword arguments inside a filter, exclusively (`FilterArguments` would then be an enum instead).

Reply:
> In my opinion, the keyword arguments should just be skipped in the positional count (so in filter: 1, a:2, 3, 3 would be the 2nd positional arg).
>
> Another thing that I've been thinking is: is it really needed to have both positional and keyword args in the same filter? Is there any liquid/jekyll/other filter that does that? Maybe we could just allow either positional or keyword arguments inside a filter, exclusively (FilterArguments would then be an enum instead).

See below about semantics

> As for the alternatives, wouldn't that make it hard to lookup the keyword arguments?

That depends on implementation.  You'd need a `match` for the keyword but it makes checking for exhaustion easy.

Reply:
I'm unsure on which of the alternatives would be better for performance:
- Having a `HashMap` where we remove every element and then assert it is empty;
- Having a `HashMap` where we call `.iter()` and use a `match` on the keyword;
- Having a `Vec<(&str, &str)>`, call `.iter()` and use a `match` on the keyword;
- Some other alternative?

Right now I'm starting to be more inclined to `Vec<(&str, &str)>`, but I'm not completely sure.

Reply:
Removing from the data structure is extra work that we don't need (and requires things to be `mut` more than needed), so I'm leaning towards `match`.

Since we don't need the other properties of a hash map and they have extra creation overhead, I'd lean towards a `Vec` also.